### DISCUSSION TITLE
Installation on Mac OSX

### DISCUSSION THREAD

**Discussion by steven-murray** [21cmfast/21cmFAST#208] [Q&A]
Several people have expressed difficulty in getting 21cmFAST installed on OSX (especially newer versions), typically due to linking to `openmp`. 

I would like to create a detailed set of installation instructions/FAQs that cover different random cases, so I'm opening this issue for people to comment on. 

@nicholebarry has posted these instructions for OSX, assuming one uses `conda`:

```
conda create -n 21cmfast python=3
conda activate 21cmfast
conda install numpy scipy jupyter matplotlib astropy h5py
conda install gsl

#OpenMP enabled libraries of fftw are available through conda-forge, using "conda install fftw" does not include openmp threads
conda install -c conda-forge fftw
conda install -c anaconda gcc

#Need to install header files specifically on macOS via Apple package in xcode developer tools. If you already have xcode developer tools, this line will error. If it fails to download, then the dmg can be downloaded from the Apple store for free.
xcode-select --install

#For older versions of xcode, use
open /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_<input version>.pkg
# and then pip install command should work

#Otherwise, for new versions of xcode, use CFLAGS="-isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX<input version>.sdk" in pip install command
```

Can these instructions be simplified at all? Do they work in all cases?

**Comment by nicholebarry** [comment]
These instructions worked for macOS Mojave 10.14.6 (the second most recent version) with the Xcode developer tools version 11 (most recent version). David (@dprelogo) said he got this working with homebrew very easily, but in general mixing conda and homebrew is a bad idea.

**Comment by dprelogo** [comment]
The only difference is I haven't used xcode's or conda's gcc, but the one form brew and passed it with CC flag.

**Comment by pritchardjr** [comment]
For me (MacOSX 10.14.6 with Xcode 11.3) this worked, but I needed CFLAGS="-isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk" rather than CFLAGS="-isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk". The latter threw errors with SDK not existing, even though the file was there.

**Comment by binnietom** [comment]
For 'MacOS Catalina 10.15.3' it worked with
$ CFLAGS="-isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk" pip install -e .

**Comment by jaguirre** [comment]
So, following @nicholebarry's instructions and creating a new environment, etc, etc, as of 8 June 2020, with Mojave 10.14.6, I found that both

`conda install -c conda-forge gcc`

and

`conda install -c anaconda gcc`

both gave 

`PackagesNotFoundError: The following packages are not available from current channels`

After doing some puzzled Googling, I found this page https://anaconda.org/conda-forge/gcc

The third option down

`conda install -c conda-forge/label/cf201901 gcc`

did result in a successful install of `gcc` and subsequently `py21cmfast`.

Oddly, running 

`conda install -c conda-forge gcc`

_again_ in this environment now succeeded.  It's possible this is all just something to do with me having something badly configured in my channels (I would have thought explicitly specifying the channel would obviate that), but another CHAMP student had a similar problem.

**Comment by steven-murray** [comment]
@jaguirre great, thanks for the update. I love how the solution for MacOS just keeps changing...

**Comment by steven-murray** [comment]
A note for all: 21cmFAST now has a `conda-forge` package! The easiest way to install should be `conda install -c conda-forge 21cmFAST`

**Reply by askielboe** [reply]
Thanks for your work! Any plans for arm64 support? 🙏 :)

**Reply by askielboe** [reply]
For reference I successfully installed 21cmFAST using `pip` following the very helpful [instructions in the documentation](https://21cmfast.readthedocs.io/en/latest/installation.html#macosx) (installing `gsl` and `fftw` using conda and setting `CC=gcc`). Thanks again!

**Reply by steven-murray** [reply]
Hi @askielboe arm64 is not something I have any experience with... would welcome PRs along those lines if you have the experience!

**Comment by steven-murray** [comment]
Closing this issue, but keeping it pinned.

**Comment by jusschwag** [comment]
Just in case it may be useful for anyone in the future, I was able to install 21cmFAST on an M4 Macbook Air with macOS 26 without conda. This is assuming you have homebrew installed and are inside the 21cmFAST directory:
```
brew install gcc
brew install fftw
brew install gsl
CC=/opt/homebrew/bin/gcc-12 CFLAGS="-I/opt/homebrew/include -L/opt/homebrew/lib" pip install -e .
```
It is important that you use gcc-12, gcc-15 will not work.

**Reply by steven-murray** [reply]
Thank you!
